velox/exec/Window.cpp

```diff
@@ -137,12 +137,58 @@ void Window::createWindowFunctions(
         operatorCtx_->pool(),
         &stringAllocator_));
 
-    windowFrames_.push_back(
+    auto updateKBasedFrameBounds = [&](const core::TypedExprPtr& frame,
+                                       WindowFrame& windowFrame,
+                                       bool isStartFrame) -> FrameChannelArg {
+      if (auto constant = constantArg(frame)) {
+        auto constantOffset =
+            constant->template as<ConstantVector<int64_t>>()->valueAt(0);
+        VELOX_USER_CHECK_GE(
+            constantOffset, 1, "k in frame bounds must be at least 1");
+        return {kConstantChannel, constant};
+      } else {
+        FlatVectorPtr<int64_t> offsets =
+            std::dynamic_pointer_cast<FlatVector<int64_t>>(
+                BaseVector::create(BIGINT(), 0, operatorCtx_->pool()));
+        return {fieldArgToChannel(frame).value(), offsets};
+      }
+    };
+
+    bool hasKBasedStartBound =
+        (windowNodeFunction.frame.startType ==
+             core::WindowNode::BoundType::kPreceding ||
+         windowNodeFunction.frame.startType ==
+             core::WindowNode::BoundType::kFollowing);
+    bool hasKBasedEndBound =
+        (windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kPreceding ||
+         windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kFollowing);
+
+    WindowFrame windowFrame(
         {windowNodeFunction.frame.type,
          windowNodeFunction.frame.startType,
          windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+         std::nullopt,
```

Comment:
The frameArgs can be constructed in place 

hasKBasedStartBound ? updateKBasedFrameBounds(....) : std::nullopt